### ISSUE TITLE
Activity cliff molecules (≥1.0 log unit, sim>0.85) degrade model performance across all endpoints with sufficient cliff populations

### ISSUE THREAD

**Issue by srijitseal** [srijitseal/polaris#12]
## Summary

Activity cliffs — pairs of structurally similar molecules (Tanimoto similarity > 0.85) with large activity differences (top quartile) — comprise 6–10% of molecules per endpoint. XGBoost trained with cluster-split CV shows higher error on cliff molecules than non-cliff molecules across all 9 endpoints, with the largest gaps on Caco-2 endpoints. The effect is moderate but consistent, confirming that smoothly interpolating models struggle where structure-activity relationships are discontinuous.

## Method

1. **Identify activity cliffs** per endpoint using the precomputed ECFP4 Tanimoto distance matrix:
   - Find all pairs with Tanimoto similarity > 0.85 (distance < 0.15)
   - Compute activity differences on log-transformed scale (except LogD)
   - Define cliff pairs as those with activity difference in the top quartile among similar pairs
   - A molecule is a "cliff molecule" if it participates in at least one cliff pair
2. **Train and evaluate** using cluster-split CV (repeat 0, 5 folds) with default XGBoost on ECFP4 + RDKit 2D descriptors
3. **Partition test predictions** into cliff vs non-cliff and compute competition metrics separately

### Cliff prevalence

| Endpoint | n molecules | n cliff pairs | n cliff mols | % cliff | diff threshold |
|----------|------------|--------------|-------------|---------|---------------|
| LogD | 7,309 | 555 | 729 | 10.0% | 0.600 |
| KSOL | 7,298 | 548 | 701 | 9.6% | 0.302 |
| Caco-2 Papp A>B | 3,773 | 246 | 377 | 10.0% | 0.202 |
| Caco-2 Efflux | 3,777 | 246 | 340 | 9.0% | 0.269 |
| HLM CLint | 4,541 | 252 | 388 | 8.5% | 0.396 |
| MLM CLint | 5,692 | 355 | 440 | 7.7% | 0.503 |
| MBPB | 1,426 | 52 | 96 | 6.7% | 0.154 |
| MPPB | 1,756 | 65 | 115 | 6.5% | 0.247 |
| MGMB | 431 | 15 | 27 | 6.3% | 0.259 |

Cliff prevalence scales with endpoint coverage — well-covered endpoints (LogD, KSOL) have more similar pairs and thus more cliffs.

### Per-endpoint performance: cliff vs non-cliff

| Endpoint | Non-cliff RAE | Cliff RAE | Non-cliff R² | Cliff R² | Non-cliff Spearman | Cliff Spearman |
|----------|-------------|-----------|-------------|---------|-------------------|---------------|
| Caco-2 Papp A>B | 0.695 | **1.050** | 0.427 | -0.187 | 0.651 | 0.329 |
| MPPB | 0.718 | 0.914 | 0.429 | 0.102 | 0.659 | 0.479 |
| Caco-2 Efflux | 0.609 | 0.819 | 0.544 | 0.195 | 0.700 | 0.580 |
| MGMB | 0.608 | 0.772 | 0.515 | 0.233 | 0.725 | 0.550 |
| KSOL | 0.724 | 0.837 | 0.386 | 0.224 | 0.554 | 0.450 |
| HLM CLint | 0.799 | 0.872 | 0.316 | 0.179 | 0.573 | 0.469 |
| MLM CLint | 0.764 | 0.685 | 0.404 | 0.489 | 0.618 | 0.728 |
| MBPB | 0.623 | 0.622 | 0.528 | 0.575 | 0.743 | 0.761 |
| LogD | 0.553 | 0.605 | 0.660 | 0.544 | 0.813 | 0.767 |

Cliff RAE exceeds non-cliff RAE for 7 of 9 endpoints. The effect is most dramatic for Caco-2 Papp A>B (cliff RAE > 1.0, meaning worse than predicting the mean) and Caco-2 Efflux. The two exceptions — MLM CLint and MBPB — show cliff molecules with *better* performance, possibly because cliff pairs in these endpoints have large value ranges that XGBoost handles via tree splitting.

R² drops on cliffs for 7 of 9 endpoints, going negative for Caco-2 Papp A>B. Spearman ρ is consistently lower on cliffs (all 9 endpoints), even for MLM CLint and MBPB — ranking cliff molecules is universally harder than ranking non-cliff molecules.

### Interpretation

The moderate but consistent cliff penalty confirms that XGBoost (a smoothly interpolating model on fingerprint features) struggles at structure-activity discontinuities. The effect is not catastrophic because:
1. Cliff molecules are the *minority* (~7–10%) in each endpoint
2. Cluster-split CV already introduces some distribution shift, raising the non-cliff baseline error
3. XGBoost's tree structure can partially capture discontinuities, unlike kernel methods

The practical takeaway: activity cliffs are a known failure mode that aggregate metrics obscure. Reporting cliff-specific performance would give practitioners a more honest view of where predictions can be trusted.

## Plots

<img width="2063" height="727" alt="Image" src="https://github.com/user-attachments/assets/30330648-b2c4-449a-a902-e6ad84fc4130" />
<img width="1478" height="727" alt="Image" src="https://github.com/user-attachments/assets/c55c6cb2-a499-4775-9def-3c2544881886" />
<img width="1478" height="727" alt="Image" src="https://github.com/user-attachments/assets/e3a650c1-e3e4-429d-8eba-5952b5e1b641" />
<img width="1478" height="727" alt="Image" src="https://github.com/user-attachments/assets/ae73139d-b3d4-4a40-b217-3cf75a220e2a" />
<img width="1478" height="727" alt="Image" src="https://github.com/user-attachments/assets/f1d5902d-6fee-47eb-92bf-8a65c7b3b74b" />
<img width="1478" height="727" alt="Image" src="https://github.com/user-attachments/assets/312d5044-a96c-4167-a3a3-98853023a646" />
<img width="1478" height="727" alt="Image" src="https://github.com/user-attachments/assets/f8ce12f5-1c81-477f-86e1-cfefcaaa9f6a" />
<img width="2678" height="1824" alt="Image" src="https://github.com/user-attachments/assets/5338d838-43cd-4d9a-899e-b4040cad7482" />

## Reproduce

```bash
pixi run -e cheminformatics python notebooks/2.10-seal-activity-cliff-eval.py
```

## Source

`notebooks/2.10-seal-activity-cliff-eval.py`

**Comment by srijitseal** [comment]
## Follow-up: sensitivity of cliff definition to the 0.85 Tanimoto threshold

A reviewer could reasonably ask whether the Tanimoto ≥ 0.85 cutoff is arbitrary. To address this, I added a `sensitivity` subcommand to `notebooks/2.10-seal-activity-cliff-eval.py` that reuses the existing out-of-fold predictions (no retraining) and sweeps the similarity threshold across {0.70, 0.75, 0.80, 0.85, 0.90, 0.95} while holding |Δactivity| ≥ 1.0 log unit fixed.

### Cliff prevalence (% of molecules involved in ≥1 cliff pair)

| Endpoint | 0.70 | 0.75 | 0.80 | **0.85** | 0.90 | 0.95 |
|---|---|---|---|---|---|---|
| LogD | 54.3% | 37.3% | 19.0% | **4.57%** | 1.00% | 0.27% |
| KSOL | 36.5% | 23.3% | 11.9% | **3.47%** | 1.16% | 0.49% |
| HLM CLint | 20.8% | 11.1% | 4.23% | **0.81%** | 0.26% | 0.09% |
| MLM CLint | 25.4% | 15.2% | 7.27% | **2.53%** | 0.74% | 0.42% |
| Caco-2 Papp A>B | 6.68% | 2.68% | 0.77% | **0.16%** | 0% | 0% |
| Caco-2 Efflux | 13.8% | 7.57% | 3.07% | **0.58%** | 0.24% | 0.05% |
| MPPB | 4.61% | 1.65% | 0.46% | **0.23%** | 0% | 0% |
| MBPB | 3.09% | 1.12% | 0.14% | **0%** | 0% | 0% |
| MGMB | 6.96% | 3.02% | 1.86% | **1.16%** | 0.46% | 0% |

Prevalence drops by ~10× between 0.70 and 0.85 and by another ~5× between 0.85 and 0.95. Four endpoints hit zero cliff pairs by 0.90; above 0.90 only LogD/KSOL/MLM/MGMB retain enough pairs to compute a cliff RAE at all.

### RAE gap (cliff RAE − non-cliff RAE)

| Endpoint | 0.70 | 0.75 | 0.80 | **0.85** | 0.90 | 0.95 |
|---|---|---|---|---|---|---|
| LogD | −0.039 | −0.017 | +0.002 | **+0.102** | +0.222 | +0.198 |
| KSOL | −0.018 | +0.014 | +0.051 | **+0.105** | +0.198 | +0.267 |
| HLM CLint | +0.096 | +0.055 | +0.085 | **+0.055** | +0.170 | n<5 |
| MLM CLint | +0.003 | +0.030 | +0.103 | **+0.148** | +0.170 | +0.211 |
| Caco-2 Papp A>B | +0.018 | +0.052 | +0.096 | **+0.264** | n=0 | n=0 |
| Caco-2 Efflux | +0.121 | +0.207 | +0.278 | **+0.368** | +0.379 | n<5 |
| MPPB | +0.036 | +0.080 | +0.243 | **n<5** | n=0 | n=0 |
| MBPB | +0.147 | +0.288 | n<5 | **n=0** | n=0 | n=0 |
| MGMB | +0.287 | +0.214 | +0.242 | **+0.212** | n<5 | n=0 |

### Interpretation: 0.85 is the empirical knee of the curve for this dataset

1. **Below 0.80 the "cliff" label is not selective.** At 0.70–0.75 the RAE gap is near-zero or even negative for LogD, KSOL, MLM CLint — meaning molecules labelled as cliffs are not measurably harder than the rest. The cliff definition only starts to isolate a genuinely harder subpopulation once sim ≥ 0.80, and the gap is reliably established at 0.85.
2. **Above 0.90 the statistics collapse.** Four of nine endpoints have zero cliff pairs at 0.90; above 0.90 only LogD and KSOL retain >20 cliff test molecules. Any "stricter" definition sacrifices statistical power without moving the qualitative conclusion.
3. **0.85 on ECFP4 is not a universal convention.** The activity-cliff literature uses a range of similarity thresholds and metrics: Stumpfe & Bajorath (2014) report 0.85 on MACCS keys (not ECFP4; the ECFP4 threshold giving the same similar-pair fraction is ~0.56) with ≥100× (≥2 log unit) potency for MMP-cliffs; van Tilborg et al. 2022 (MoleculeACE) use ≥0.9 under a soft consensus across ECFP4/scaffold/SMILES similarities with ≥10× (≥1 log unit) potency. Our 0.85 on ECFP4 with ≥1 log unit is a distinct operating point, and the sensitivity analysis above justifies it on empirical grounds specific to this dataset: at thresholds lower than ~0.80 the cliff label is not selective, and above ~0.90 the per-endpoint cliff populations are too small for reliable RAE estimation.

### Outputs saved

`data/processed/2.10-seal-activity-cliff-eval/`
- `cliff_sensitivity.csv` — 54 rows (9 endpoints × 6 thresholds) with cliff counts, cliff/non-cliff RAE, and RAE gap
- `cliff_sensitivity_pairs.parquet` — 118,773 similar pairs (sim ≥ 0.70) for downstream inspection
- `cliff_sensitivity_pct.png` — % cliff molecules vs threshold (one panel per endpoint)
- `cliff_sensitivity_rae_gap.png` — RAE gap vs threshold (one panel per endpoint)
- `cliff_sensitivity_joint.png` — hexbin of (similarity, |Δactivity|) per endpoint with cliff-definition lines overlaid at (0.85, 1.0)

### Reproduce

```bash
pixi run -e cheminformatics python notebooks/2.10-seal-activity-cliff-eval.py sensitivity
```

The subcommand reuses `cliff_vs_noncliff_errors.csv` from the `main` command — no model retraining.

---

<img width="2215" height="1804" alt="Image" src="https://github.com/user-attachments/assets/72ba5118-b443-4287-b1ac-fa59b4845a67" />
<img width="1325" height="878" alt="Image" src="https://github.com/user-attachments/assets/816efaf7-7f93-4d14-970b-9501bc738822" />
<img width="1324" height="878" alt="Image" src="https://github.com/user-attachments/assets/2a10d334-1294-4917-b294-ddb005061d42" />

**Comment by srijitseal** [comment]
## CheMeleon update (2.10 activity cliffs)

Re-ran the cliff evaluation with CheMeleon (cluster-split CV, 5 folds, fine-tuned per fold). Cliff definition unchanged: Tanimoto similarity > 0.85 + |Δactivity| ≥ 1.0 log unit. Cliff/non-cliff partitions are identical (counts unchanged).

**Verdict: trend matches XGBoost. Cliff RAE > non-cliff RAE for all 7 endpoints with sufficient cliffs.**

### Cliff RAE vs non-cliff RAE

| Endpoint | n_cliff | RAE cliff XGB | RAE non-cliff XGB | RAE cliff CM | RAE non-cliff CM |
|----------|---------|---------------|-------------------|--------------|------------------|
| LogD | 334 | 0.592 | 0.491 | 0.422 | 0.360 |
| KSOL | 253 | 0.807 | 0.698 | 0.713 | 0.617 |
| HLM CLint | 37 | 0.865 | 0.810 | 0.939 | 0.792 |
| MLM CLint | 144 | 0.829 | 0.704 | 0.801 | 0.684 |
| Caco-2 Papp A>B | 6 | 0.879 | 0.650 | 0.942 | 0.677 |
| Caco-2 Efflux | 22 | 0.962 | 0.585 | 0.895 | 0.586 |
| MGMB | 5 | 0.776 | 0.555 | 0.944 | 0.501 |

MPPB (4 cliffs) and MBPB (0 cliffs) excluded from cliff evaluation in both models.

### Cliff R²

| Endpoint | R² cliff XGB | R² cliff CM |
|----------|--------------|-------------|
| LogD | 0.584 | 0.777 |
| KSOL | 0.294 | 0.274 |
| HLM CLint | 0.240 | 0.150 |
| MLM CLint | 0.302 | 0.374 |
| Caco-2 Papp | 0.109 | −0.256 |
| Caco-2 Efflux | −0.006 | 0.075 |
| MGMB | −0.226 | −0.701 |

### Trend match

Yes, fully. The headline finding — *cliff molecules are systematically harder to predict than non-cliff molecules across endpoints with sufficient cliff populations* — replicates verbatim. The cliff RAE is greater than non-cliff RAE for **7/7 endpoints with sufficient cliffs** in both models.

CheMeleon does *not* selectively help cliff molecules — non-cliff RAE drops by 0.05–0.13 from XGBoost, while the cliff/non-cliff RAE gap remains 0.06–0.44 (median ~0.13). The interpolation failure mode is preserved under a foundation model.

Caco-2 Papp R² actually goes more negative under CheMeleon (+0.11 → −0.26), and MGMB cliff R² goes from −0.23 to −0.70 — but cliff sample sizes (6 and 5) are too small for these to be load-bearing.

### Source

- `data/processed/2.10-seal-activity-cliff-eval/chemeleon/summary_metrics.csv`
- `data/processed/2.10-seal-activity-cliff-eval/combined/`
- Reproduce: `pixi run -e cheminformatics python notebooks/2.10-seal-activity-cliff-eval.py main --model chemeleon`

**Comment by srijitseal** [comment]
## Refreshed numbers (XGBoost rerun + CheMeleon)

Issue body re-synced against `data/processed/2.10-seal-activity-cliff-eval/xgboost/`. Cliff RAE table drift (cliff/non-cliff RAE):

| Endpoint | Old | Refreshed |
|----------|-----|-----------|
| Caco-2 Efflux | 0.949 / 0.581 | **0.962 / 0.585** |
| Caco-2 Papp | 0.907 / 0.643 | **0.879 / 0.650** |
| HLM CLint | 0.866 / 0.811 | **0.865 / 0.810** |
| MLM CLint | 0.843 / 0.696 | **0.829 / 0.704** |
| KSOL | 0.807 / 0.703 | **0.807 / 0.698** |
| MGMB | 0.760 / 0.548 | **0.776 / 0.555** |
| LogD | 0.593 / 0.491 | **0.592 / 0.491** |

Cliff R² also refreshed; cliff R² is now negative for Caco-2 Efflux (−0.006) and MGMB (−0.226). Sensitivity-sweep prose updated: prevalence drop 0.70→0.85 is **~12.7×**, 0.85→0.95 is **~10.2×**; 3 of 9 endpoints have zero cliff pairs at 0.90, rising to 4 of 9 at 0.95. ADR link path corrected (`../../decisions/...`); reproduce command updated to include the required `main` subcommand.

CheMeleon comparison (cliff RAE > non-cliff RAE for all 7 endpoints with sufficient cliffs) stands — see #issuecomment-4381083314.

**Comment by srijitseal** [comment]
**Update: Combined figure created**

Combined Figure S4 (cliff_sensitivity_combined.png) added to the combined/ folder. Layout is now vertical with three panels: (a) prevalence bar chart, (b) RAE gap for both XGBoost and CheMeleon overlaid, (c) hexbin density. XGBoost and CheMeleon results are shown side-by-side within each panel.
